core/chains/evm/abi/selector_parser.go

```diff
@@ -0,0 +1,249 @@
+}
+
+func parseElementaryType(unescapedSelector string) (string, string, error) {
+	parsedType, rest, err := parseToken(unescapedSelector, false)
+	if err != nil {
+		return "", "", fmt.Errorf("failed to parse elementary type: %v", err)
+	}
+	// handle arrays
+	for len(rest) > 0 && rest[0] == '[' {
+		parsedType = parsedType + string(rest[0])
+		rest = rest[1:]
+		for len(rest) > 0 && isDigit(rest[0]) {
+			parsedType = parsedType + string(rest[0])
+			rest = rest[1:]
+		}
+		if len(rest) == 0 || rest[0] != ']' {
+			return "", "", fmt.Errorf("failed to parse array: expected ']', got %c", unescapedSelector[0])
+		}
+		parsedType = parsedType + string(rest[0])
+		rest = rest[1:]
+	}
+	return parsedType, rest, nil
+}
+
+func parseCompositeType(unescapedSelector string) ([]interface{}, string, error) {
+	if len(unescapedSelector) == 0 || unescapedSelector[0] != '(' {
+		return nil, "", fmt.Errorf("expected '(', got %c", unescapedSelector[0])
+	}
+	parsedType, rest, err := parseType(unescapedSelector[1:])
+	if err != nil {
+		return nil, "", fmt.Errorf("failed to parse type: %v", err)
+	}
+	result := []interface{}{parsedType}
+	for len(rest) > 0 && rest[0] != ')' {
+		parsedType, rest, err = parseType(rest[1:])
+		if err != nil {
+			return nil, "", fmt.Errorf("failed to parse type: %v", err)
+		}
+		result = append(result, parsedType)
+	}
+	if len(rest) == 0 || rest[0] != ')' {
+		return nil, "", fmt.Errorf("expected ')', got '%s'", rest)
+	}
+	if len(rest) >= 3 && rest[1] == '[' && rest[2] == ']' {
+		return append(result, "[]"), rest[3:], nil
+	}
+	return result, rest[1:], nil
+}
+
+func parseType(unescapedSelector string) (interface{}, string, error) {
+	if len(unescapedSelector) == 0 {
+		return nil, "", errors.New("empty type")
+	}
+	if unescapedSelector[0] == '(' {
+		return parseCompositeType(unescapedSelector)
+	}
+	return parseElementaryType(unescapedSelector)
+}
+
+func parseArgs(unescapedSelector string) ([]abi.ArgumentMarshaling, error) {
+	if len(unescapedSelector) == 0 || unescapedSelector[0] != '(' {
+		return nil, fmt.Errorf("expected '(', got %c", unescapedSelector[0])
+	}
+	result := []abi.ArgumentMarshaling{}
+	rest := unescapedSelector[1:]
+	var parsedType any
+	var err error
+	for len(rest) > 0 && rest[0] != ')' {
+		// parse method name
+		var name string
+		name, rest, err = parseIdentifier(rest[:])
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse name: %v", err)
+		}
+
+		// skip whitespace between name and identifier
+		for rest[0] == ' ' {
+			rest = rest[1:]
+		}
+
+		// parse type
+		parsedType, rest, err = parseType(rest[:])
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse type: %v", err)
+		}
+
+		arg, err := assembleArg(name, parsedType)
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse type: %v", err)
+		}
+
+		result = append(result, arg)
+
+		for rest[0] == ' ' || rest[0] == ',' {
+			rest = rest[1:]
+		}
+	}
+	if len(rest) == 0 || rest[0] != ')' {
```

Comment:
The `if` statements here are a little awkward. Not going to necessarily request changes, but I would suggest something simpler like `if !(len(rest) == 1 && rest[0] == ')')`